### ISSUE TITLE
-prod option fails with cryptic error since upgrade to >0.0.34

### ISSUE THREAD

**Issue by sgbeal** [angular/angular-cli#700]
1. OS. Windows 7, 8 or 10. Linux (which distribution). Mac OSX (Yosemite? El Capitan?)

- Windows 10. All consoles are run in admin mode.
- Mint Linux 17.3 on x64

2. Versions. Please run `ng --version`. If there's nothing outputted, please
   run in a Terminal: 

        $ ng --version
        (node:7172) fs: re-evaluating native module sources is not supported. If you are using the graceful-fs module, please update it to a more recent version.
        angular-cli: 1.0.0-beta.1
        node: 6.1.0
        os: win32 x64

Plus, for completeness...

	$ node --version; npm --version; npm info angular-cli
	v6.1.0
	3.8.6

	{ name: 'angular-cli',
	  time:
	   { modified: '2016-05-07T16:07:33.383Z',
		 created: '2014-10-07T04:34:38.078Z',
		 '1.0.0': '2014-10-07T04:34:38.078Z',
		 '0.0.0': '2014-10-11T11:38:24.914Z',
		 '0.0.1': '2015-08-03T17:12:04.553Z',
		 '0.0.2': '2015-10-18T17:46:40.146Z',
		 '0.0.3': '2015-10-18T21:42:52.445Z',
		 '0.0.4': '2015-10-19T01:14:13.216Z',
		 '0.0.5': '2015-10-19T09:14:01.712Z',
		 '0.0.6': '2015-10-19T09:36:51.644Z',
		 '0.0.7': '2015-10-19T11:48:24.233Z',
		 '0.0.8': '2015-10-19T11:51:59.318Z',
		 '0.0.9': '2015-10-19T21:18:13.725Z',
		 '0.0.10': '2015-10-19T22:18:13.091Z',
		 '0.0.11': '2015-11-03T05:03:06.345Z',
		 '0.0.12': '2015-11-15T10:10:48.551Z',
		 '0.0.13': '2015-12-04T17:16:43.662Z',
		 '0.0.14': '2015-12-05T01:40:39.377Z',
		 '0.0.15': '2015-12-05T05:50:59.226Z',
		 '0.0.16': '2015-12-10T12:23:28.722Z',
		 '0.0.17': '2015-12-14T19:37:08.400Z',
		 '0.0.18': '2015-12-15T12:51:47.990Z',
		 '0.0.19': '2015-12-15T18:23:33.077Z',
		 '0.0.20': '2015-12-15T20:39:55.357Z',
		 '0.0.22': '2016-02-20T00:36:56.908Z',
		 '0.0.23': '2016-02-20T01:44:13.897Z',
		 '0.0.24': '2016-02-25T17:41:20.619Z',
		 '0.0.25': '2016-03-10T02:10:05.410Z',
		 '0.0.26': '2016-03-23T00:39:46.402Z',
		 '0.0.27': '2016-03-23T22:05:56.538Z',
		 '0.0.28': '2016-03-28T22:39:45.766Z',
		 '0.0.29': '2016-03-30T18:10:47.060Z',
		 '0.0.30': '2016-04-06T22:27:32.015Z',
		 '0.0.31': '2016-04-11T18:20:32.677Z',
		 '0.0.33': '2016-04-19T22:31:48.209Z',
		 '0.0.34': '2016-04-26T17:56:20.524Z',
		 '0.0.37': '2016-05-03T18:49:21.731Z',
		 '0.0.39': '2016-05-04T16:29:23.054Z',
		 '0.1.0': '2016-05-05T19:41:05.830Z',
		 '1.0.0-beta.0': '2016-05-06T18:18:54.950Z',
		 '1.0.0-beta.1': '2016-05-07T16:07:33.383Z' },
	  maintainers:
	   [ 'angularcore <angular-core+npm@google.com>',
		 'rodyhaddad <npm@rodyhaddad.com>' ],
	  description: 'CLI tool for Angular',
	  versions:
	   [ '0.0.0',
		 '0.0.1',
		 '0.0.2',
		 '0.0.3',
		 '0.0.4',
		 '0.0.5',
		 '0.0.6',
		 '0.0.7',
		 '0.0.8',
		 '0.0.9',
		 '0.0.10',
		 '0.0.11',
		 '0.0.12',
		 '0.0.13',
		 '0.0.14',
		 '0.0.15',
		 '0.0.16',
		 '0.0.17',
		 '0.0.18',
		 '0.0.19',
		 '0.0.20',
		 '0.0.22',
		 '0.0.23',
		 '0.0.24',
		 '0.0.25',
		 '0.0.26',
		 '0.0.27',
		 '0.0.28',
		 '0.0.29',
		 '0.0.30',
		 '0.0.31',
		 '0.0.33',
		 '0.0.34',
		 '0.0.37',
		 '0.0.39',
		 '0.1.0',
		 '1.0.0-beta.0',
		 '1.0.0-beta.1' ],
	  'dist-tags': { latest: '1.0.0-beta.1', beta: '1.0.0-beta.1' },
	  readmeFilename: 'README.md',
	  homepage: 'https://github.com/angular/angular-cli',
	  keywords: [],
	  repository:
	   { type: 'git',
		 url: 'git+https://github.com/angular/angular-cli.git' },
	  author: 'Angular Authors',
	  bugs: { url: 'https://github.com/angular/angular-cli/issues' },
	  license: 'MIT',
	  contributors:
	   [ 'Rody Haddad <npm@rodyhaddad.com> (http://rodyhaddad.com/)',
		 'Igor Minar <igor@angularjs.org>' ],
	  users: { altrim: true, seachange: true },
	  version: '1.0.0-beta.1',
	  main: 'lib/cli/index.js',
	  trackingCode: 'UA-8594346-19',
	  bin: { ng: './bin/ng' },
	  scripts: { test: 'node tests/runner', lint: 'eslint .' },
	  engines: { node: '>= 4.1.0' },
	  dependencies:
	   { broccoli: '^1.0.0-beta.7',
		 'broccoli-caching-writer': '^2.2.1',
		 'broccoli-concat': '^2.2.0',
		 'broccoli-funnel': '^1.0.1',
		 'broccoli-merge-trees': '^1.1.1',
		 'broccoli-source': '^1.1.0',
		 'broccoli-writer': '^0.1.1',
		 chalk: '^1.1.3',
		 'ember-cli': '2.5.0',
		 'ember-cli-string-utils': '^1.0.0',
		 exit: '^0.1.2',
		 'fs-extra': '^0.30.0',
		 glob: '^7.0.3',
		 leek: '0.0.21',
		 lodash: '^4.11.1',
		 opn: '4.0.1',
		 resolve: '^1.1.7',
		 shelljs: '^0.7.0',
		 'silent-error': '^1.0.0',
		 'symlink-or-copy': '^1.0.3',
		 'systemjs-builder': '^0.15.16',
		 typescript: '^1.8.10',
		 typings: '^0.8.1' },
	  'ember-addon': { paths: [ './addon/ng2/' ] },
	  devDependencies:
	   { chai: '^3.5.0',
		 'clang-format': '^1.0.38',
		 eslint: '^2.8.0',
		 'exists-sync': '0.0.3',
		 minimatch: '^3.0.0',
		 mocha: '^2.4.5',
		 'mock-fs': '^3.8.0',
		 'object-assign': '^4.0.1',
		 rewire: '^2.5.1',
		 sinon: '^1.17.3',
		 through: '^2.3.8',
		 tslint: '^3.8.1',
		 'walk-sync': '^0.2.6' },
	  gitHead: '668ecee86bc9acbf5fc41994802eb263b8b9c4aa',
	  dist:
	   { shasum: '86feceee719d7a14579ab4486ca3077a1a80a840',
		 tarball: 'https://registry.npmjs.org/angular-cli/-/angular-cli-1.0.0-beta.1.tgz' },
	  directories: {} }

3. Repro steps. Tell us exactly what you did and how you came to the failure. 

after upgrading our project to angular2 rc.1 and angular-cli 1.0.0-beta, the -prod option no longer works. We get a completely mysterious error message which does not give us enough information to even guess at what the problem is:

	$ ng server --prod
	(node:7152) fs: re-evaluating native module sources is not supported. If you are using the graceful-fs module, please update it to a more recent version.
	Livereload server on http://localhost:49152
	Serving on http://localhost:4200/
	(node:7152) DeprecationWarning: 'root' is deprecated, use 'global'
	(node:7152) DeprecationWarning: 'GLOBAL' is deprecated, use 'global'
	The Broccoli Plugin: [BundlePlugin] failed with:
	Error on fetch for main at file:///D:/<elided>/nms-gui/tmp/bundle_plugin-input_base_path-ZohVXdOM.tmp/0/main
			Error: ENOENT: no such file or directory, open 'D:\<elided>\nms-gui\tmp\bundle_plugin-input_base_path-ZohVXdOM.tmp\0\main'
		at Error (native)

	The broccoli plugin was instantiated at:
		at BundlePlugin.Plugin (D:\<elided>\nms-gui\node_modules\broccoli-plugin\index.js:10:31)
		at BundlePlugin.CachingWriter [as constructor] (D:\<elided>\nms-gui\node_modules\broccoli-caching-writer\index.js:21:10)
		at BundlePlugin (D:\<elided>\nms-gui\node_modules\angular-cli\lib\broccoli\angular-broccoli-bundle.js:11:5)
		at Angular2App._getBundleTree (D:\<elided>\nms-gui\node_modules\angular-cli\lib\broccoli\angular2-app.js:375:55)
		at Angular2App._buildTree (D:\<elided>\nms-gui\node_modules\angular-cli\lib\broccoli\angular2-app.js:137:21)
		at new Angular2App (D:\<elided>\nms-gui\node_modules\angular-cli\lib\broccoli\angular2-app.js:42:23)
		at module.exports (D:\<elided>\nms-gui\angular-cli-build.js:6:10)
		at Class.module.exports.Task.extend.setupBroccoliBuilder (D:\<elided>\nms-gui\node_modules\angular-cli\lib\models\builder.js:55:19)
		at Class.module.exports.Task.extend.init (D:\<elided>\nms-gui\node_modules\angular-cli\lib\models\builder.js:89:10)
		at new Class (D:\<elided>\nms-gui\node_modules\core-object\core-object.js:18:12)
		at Class.module.exports.Task.extend.run (D:\<elided>\nms-gui\node_modules\angular-cli\lib\tasks\serve.js:15:19)
		at D:\<elided>\nms-gui\node_modules\angular-cli\lib\commands\serve.js:64:24
		at lib$rsvp$$internal$$tryCatch (D:\<elided>\nms-gui\node_modules\rsvp\dist\rsvp.js:1036:16)
		at lib$rsvp$$internal$$invokeCallback (D:\<elided>\nms-gui\node_modules\rsvp\dist\rsvp.js:1048:17)
		at lib$rsvp$$internal$$publish (D:\<elided>\nms-gui\node_modules\rsvp\dist\rsvp.js:1019:11)
		at lib$rsvp$asap$$flush (D:\<elided>\nms-gui\node_modules\rsvp\dist\rsvp.js:1198:9)

(We see the same with node 4.4.3 and 5.x)

All of the classes referenced there come from/via a2 and/or angular-cli - none of them are from our tree.

   If you can give us steps from the moment you created an empty application,
   even better!

Unfortunately this does not happen with a clean-slate (ng new) app, but:

a) we updated our application by using (ng new) and copying our sources over to the new tree.

b) non-prod builds and runs just fine, so we have to assume the problem is somewhere internally in the handling of the prod environment (which we do not manipulate, other than to add 2 empty functions to the environment object in config/environment.prod.ts). Unfortunately, ng does not support custom environment names (`--environment=mine`), so attempts to reproduce it (or rule it out) that way have failed.

c) we've gone through and compared our config files with those of a clean-slate build literally 5 times, and the only difference is that our app  lives under src/client/app instead of src/app (i'm told by our maven gurus that that's necessary for our particular build environment).

Because the error message is so cryptic (and contains no mention of classes/files from us), we have literally no idea where to start looking. Any hint for how we can tickle more info out of -prod would be much appreciated.

**Comment by sgbeal** [comment]
BTW: this affects all uses of -prod, e.g. (ng build -prod) throws the same error.

**Comment by sgbeal** [comment]
From the stacktrace:

    ... angular-cli-build.js:6:10

looks like:

    $ sed.exe -n 5,7p angular-cli-build.js
    module.exports = function(defaults) {
      return new Angular2App(defaults, {
        vendorNpmFiles: [

just for kicks, i removed all vendorNpmFiles entries, but still the same error. i'm just shooting in the dark here, as i haven't the faintest idea what the problem is and this problem is (if i can't solve it in the next 90 minutes) going to force us to go back to the Angular2 Beta so that we can complete a customer demo by Friday.

**Comment by sgbeal** [comment]
Found something...

if i downgrade both my global and packages.json angular.cli to 0.0.34, it works as expected. Anything higher than 0.0.34 exhibits the problem described above.

**Comment by sgbeal** [comment]
Another data point:

the global version of angular-cli seems to be irrelevant! The in-tree (packages.json) copy has to be no newer than 0.0.34. The global one (from which (ng build -prod) is started) may be 1.0.0-beta1.

From outside my tree:

$ ng --version
(node:3292) fs: re-evaluating native module sources is not supported. If you are using the graceful-fs module, please update it to a more recent version.
angular-cli: 1.0.0-beta.1

from inside:
$ ng --version
(node:4228) fs: re-evaluating native module sources is not supported. If you are using the graceful-fs module, please update it to a more recent version.
angular-cli: 0.0.34
node: 6.1.0
os: win32 x64


the -prod option works in this configuration, but not if the in-tree copy is newer than 0.0.34.

**Comment by shrop** [comment]
Seeing the same issue after `npm install -g angcli` in OS X 10.11.4.

**Comment by sgbeal** [comment]
@shrop angcli appears to be a different project (http://ngcli.github.io/). This one's npm name is `angular-cli`.

**Comment by hansl** [comment]
`angcli` is a different project, totally unrelated.

`vendorNpmFiles` contains all the vendor files with version >= 0.0.35. Try including the following vendor files:

```javascript
    vendorNpmFiles: [
      'systemjs/dist/system-polyfills.js',
      'systemjs/dist/system.src.js',
      'zone.js/dist/*.js',
      'es6-shim/es6-shim.js',
      'reflect-metadata/*.js',
      'rxjs/**/*.js',
      '@angular/**/*.js'
    ]

```

**Comment by sgbeal** [comment]
Our vendorNpmFiles already contains all of those:

		vendorNpmFiles: [
		  'systemjs/dist/system-polyfills.js',
		  'systemjs/dist/system.src.js',
		  'zone.js/dist/*.js',
		  'es6-shim/es6-shim.js',
		  'reflect-metadata/*.js',
		  'rxjs/**/*.js',
		  '@angular/**/*.js',

		  'bootstrap/dist/js/bootstrap.min.js',
		  'jquery/dist/jquery.min.js',
		  'bootstrap/dist/css/bootstrap.min.css',
		  'bootstrap/dist/fonts/glyphicons-halflings-regular.eot',
		  'bootstrap/dist/fonts/glyphicons-halflings-regular.svg',
		  'bootstrap/dist/fonts/glyphicons-halflings-regular.ttf',
		  'bootstrap/dist/fonts/glyphicons-halflings-regular.woff',
		  'bootstrap/dist/fonts/glyphicons-halflings-regular.woff2',
		  'ng2-bs3-modal/bundles/ng2-bs3-modal.min.js'
		]

just for kicks, i reduced it to the bare minimum set (up to and including @angular), but the same error.

		$ ng build -prod
		(node:6116) fs: re-evaluating native module sources is not supported. If you are using the graceful-fs module, please update it to a more recent version.
		(node:6116) DeprecationWarning: 'root' is deprecated, use 'global'
		(node:6116) DeprecationWarning: 'GLOBAL' is deprecated, use 'global'
		Build failed.
		The Broccoli Plugin: [BundlePlugin] failed with:
		Error on fetch for main at file:///D:/src/x/tmp/bundle_plugin-input_base_path-VniKYUhB.tmp/0/main
				Error: ENOENT: no such file or directory, open 'D:\src\x\tmp\bundle_plugin-input_base_path-VniKYUhB.tmp\0\main'
			at Error (native)

this is using a fresh copy of node_modules, ./tmp, and ./dist, reproduced on Win10 (w/ admin rights) and Mint Linux (as normal user), with both node 4.x and 6.1.

i'm at an utter loss on how to determine which component is causing this.

**Comment by shrop** [comment]
@sgbeal thanks for clearing that up. :)

**Comment by filipesilva** [comment]
@sgbeal this seems to the bundling specifically, which follows file imports. It complains of a missing `main.js` (transpiled from `main.ts`) at the root - did you remove it in your project?

**Comment by sgbeal** [comment]
No - we still have the main.js (and bootstrap our app from there), but we have it 1 directory level deeper (src/client/app instead of src/app). Our angular-cli.js[on?] is configured for that and a non-prod builds fine. i also tried, at some point, to move our app up to src/app, to rule out that that was the problem, but the result was the same.

i won't have access to the source again until tomorrow morning (about 13 hours from now), but can post any configuration snippets you care to see.

**Comment by filipesilva** [comment]
Can I ask you then to make a repo containing everything but the content of `app/` (perhaps with a dummy `*AppComponent`) as is on your project? 

That way I should be able to clone, install and reproduce the issue.

**Comment by sgbeal** [comment]
i will certainly _try_ but i can make no guarantees because i'm currently disabled (reduced to typing with 1 hand due to nerve damage caused by too much programming), so my capacity for experimentation is sorely reduced. We've unfortunately got client-specific/secret bits which i can't publicly post, and stripping it down might require more effort than i can physically commit to. But i will commit to trying.

**Incidentally... this may turn out to be irrelevant** because we discovered today that we cannot ship systemjs-generated JS bundles with angular-cli due to limitations of angular-cli:

- ng provides no hooks to add a generation step to our build process. It simply copies/transpiles from src to dist (and node_modules to dist, but that's all 3rd-party code).

- ng is _hard-coded_ to exclude `**/*.js` when copying to ./dist, so we cannot put static/pre-generated bundles in our tree and have ng copy them. It only knows how to copy *.js from ./node_modules.

- ng refuses to serve any files it did not place in ./dist (it returns a 404 for them), so we cannot simply copy generated bundle files into dist.

- we tried adding a dummy-bundle.ts to the tree, so that ng would transpile it to dummy-bundle.js, which we would then _overwrite_ in a separate step with the contents of our bundle, but ng apparently caches the _content_ of files it copies/serves, and always delivers us the empty contents of the dummy file, not the contents we overwrite it with.

i.e. bundling is seemingly _impossible_ with angular-cli, and the recent changes in a2 which cause it to send over 300 XHR requests to start up the hello-world app (and almost 700 for our app, with only about 30 client-side classes), mean that client-side bundling has become a hard requirement (and the word from the a2 team is (paraphrased) "it's your client-side problem now").

Summary: the combination of a2's recently removal of bundles, combined with ng's inability to integrate client-generated bundles, will likely force us to revert to gulp for build process. But that remains to be seen - we're still trying to shoehorn in our bundles to ng because we like it better (there are fewer knobs to fiddle with).

**Comment by filipesilva** [comment]
Oh but that's the thing - `ng build -prod` now will bundle all of your app *and* dependencies. 

There's something going wrong with the bundle process in your upgraded project thought, that doesn't seem to happen in new projects.

Don't worry too much about editing code for the example - just remove it all. The process up to the error only needs main.ts so all of your app can be deleted from the example.

**Comment by sgbeal** [comment]
THAT i can commit to. i'll get it ready tomorrow (Central Euro. Time). However, i don't think it will reproduce that way. Week before last i tried (twice), chunk by chunk, to re-construct the source tree to find out which file/config was triggering it. A bare-bones main.js, including our base-most services and component, did not trigger it. It wasn't until i had to add a large chunk of things at once (due to how the cross-wiring/deps worked out) that it failed, and it wasn't feasible to tear that chunk down into smaller parts. i'll try it again and feed back on that tomorrow.

**Comment by sgbeal** [comment]
@filipesilva i've got a small reproduce without all the app code (just a main plus all the config bits and potentially some other cruft/temp files outside the src dir) - see attached, and then:

      unzip ng-700.zip
      cd ng-700
      npm install
      ng build -prod

that's using ng 1.0.0-beta.1.

Sidebar: you may notice a _directory_ named src/main - i tried removing that, but it didn't solve the problem (i figured the filename resolver might have been choking on it). That structure is part of our overlying maven build (which integrates the a2 content into a separate webapp) and our maven guru insists that this structure simplifies his job.

Results:
		$ ng build -prod
		(node:9536) fs: re-evaluating native module sources is not supported. If you are using the graceful-fs module, please update it to a more recent version.
		"{{content-for}}" has been deprecated and will be removed before RC.
		(node:9536) DeprecationWarning: 'root' is deprecated, use 'global'
		(node:9536) DeprecationWarning: 'GLOBAL' is deprecated, use 'global'
		Build failed.
		The Broccoli Plugin: [BundlePlugin] failed with:
		Error on fetch for main.js at file:///D:/src/ng-700/tmp/bundle_plugin-input_base_path-NvrPJzmf.tmp/0/main.js
				Error: ENOENT: no such file or directory, open 'D:\src\ng-700\tmp\bundle_plugin-input_base_path-NvrPJzmf.tmp\0\main.js'
			at Error (native)

		The broccoli plugin was instantiated at:
			at BundlePlugin.Plugin (D:\src\ng-700\node_modules\broccoli-plugin\index.js:10:31)
			at BundlePlugin.CachingWriter [as constructor] (D:\src\ng-700\node_modules\broccoli-caching-writer\index.js:21:10)
			at BundlePlugin (D:\src\ng-700\node_modules\angular-cli\lib\broccoli\angular-broccoli-bundle.js:11:5)
			at Angular2App._getBundleTree (D:\src\ng-700\node_modules\angular-cli\lib\broccoli\angular2-app.js:424:22)
			at Angular2App._buildTree (D:\src\ng-700\node_modules\angular-cli\lib\broccoli\angular2-app.js:159:21)
			at new Angular2App (D:\src\ng-700\node_modules\angular-cli\lib\broccoli\angular2-app.js:53:23)
			at module.exports (D:\src\ng-700\angular-cli-build.js:6:10)
			at Class.module.exports.Task.extend.setupBroccoliBuilder (D:\src\ng-700\node_modules\angular-cli\lib\models\builder.js:55:19)
			at Class.module.exports.Task.extend.init (D:\src\ng-700\node_modules\angular-cli\lib\models\builder.js:89:10)
			at new Class (D:\src\ng-700\node_modules\core-object\core-object.js:18:12)
			at Class.module.exports.Task.extend.run (D:\src\ng-700\node_modules\angular-cli\lib\tasks\build.js:15:19)
			at D:\src\ng-700\node_modules\angular-cli\lib\commands\build.js:32:24
			at lib$rsvp$$internal$$tryCatch (D:\src\ng-700\node_modules\rsvp\dist\rsvp.js:1036:16)
			at lib$rsvp$$internal$$invokeCallback (D:\src\ng-700\node_modules\rsvp\dist\rsvp.js:1048:17)
			at lib$rsvp$$internal$$publish (D:\src\ng-700\node_modules\rsvp\dist\rsvp.js:1019:11)
			at lib$rsvp$asap$$flush (D:\src\ng-700\node_modules\rsvp\dist\rsvp.js:1198:9)
[ng-700.zip](https://github.com/angular/angular-cli/files/269721/ng-700.zip)


if i backrev to ng 0.0.34 (in package.json), npm install, and try again:

		$ ng build -prod
		(node:1244) fs: re-evaluating native module sources is not supported. If you are using the graceful-fs module, please update it to a more recent version.
		BuildingBuilding.Built project successfully. Stored in "dist/".


**i "really, really, really, ridiculously" appreciate you looking into this!**

**Comment by filipesilva** [comment]
@sgbeal you're most welcome, I build this feature and if you're having problems with it then I want to help.

I've pinpointed the problem: you still have `main.ts`, but it has been moved inside `app`. Moving it outside and updating the SystemJS mapping from `'app/main': 'app/main.js',` to 'main': 'main.js', fixes the issue.

This is because the bundler right now is looking for *very* specific packages (`main` and `app`) and putting the generated bundle in *very* specific locations (`/main.js` and `app/index.js`), as a matter of convention.

This process can be improved to use the declared SystemJS package `main` property instead of assuming the convention-based paths, but having a package named `main` itself is something that will likely remain as a convention.

Meanwhile the solution above should help - tell me if it doesn't and I'll find another way.

**Comment by sgbeal** [comment]
Excitedly trying it out now,  but npm is barfing on me ("maximum call stack size exceeded") for other reasons.

While i'm waiting on npm to do its thing... i mentioned earlier that main.ts had been moved. The curious thing is that it works fine with versions >0.0.34 in dev builds, just not prod. Why would a prod build behave differently in this regard? (Sounds like a bug to me!)

npm done...

now i'm getting the same error but for app.js instead of main.js, but at least i now have a starting point, so... it seems something to do with us not having an app/index.js. Right... when i add an empty app/index.ts, -prod mode compiles :).

Is having an empty app/index.ts bad, or should we move the main.ts bits into app/index.ts? Looking at a (ng new) app, having both main.ts and app/index.ts seems redundant to me. i'm not sure why it's even trying to load app/index.js - none of our imports reference 'app' without a suffix, e.g. 'app/foo/bar'.

The current state is that in prod mode it compiles, but does not serve properly because it tries to dynamically load our own classes and the environment.js from angular-cli and cannot find them (presumably they're bundled). This may be an incompatibility in our systemjs config - i will need some time to weed through it.

Curiously... in non-prod mode it is still building/serving just fine.

Thank you for the tip!!! i will feed back if/when i can resolve this (hopefully tomorrow - my work day is about done already).

**Comment by filipesilva** [comment]
Apologies, I didn't see the bit about `main.ts` having been moved. 

The discrepancy between dev and prod builds is because bundling (introduced in 0.0.34 most likely) needs to be aimed at something, an entry point, whereas dev builds simply compile using tsc (which currently has a few files listed). So we choose the style guide recommendation as the entry point: `main.ts`.

Having `index.ts` re-exporting the contents of each folder is also a style guide recommendation: https://angular.io/docs/ts/latest/guide/style-guide.html#!#create-and-import-barrels

Bundling at the moment doesn't support not having `index.ts` for `app`, but that can be added really, since that information is in the SystemJS configuration as the main file for that package (while currently hardcoded in the bundling process).

I think your current issue is because you're not using the barrels actually. Essentially, the `main` bundle (that contains `main.ts` and all vendor code) tries to load the `app` bundle via SystemJS, which sees that the main file is `index.ts` - but yours is empty. 

The purpose of those `index.ts` is to serve as the package entry point. So to speak, anything that `main.ts` needs from `app/` should be obtained via `index.ts` via reexporting. In the style guide they explain it a bit better.

I can modify bundling to be more robust and read entry points from the SystemJS config, but only later this week at best, and then it would be in the next release. For now, I think you do need the barrel `index.ts` file.

**Comment by sgbeal** [comment]
Thanks for the tips and link - i will try that out first thing in the morning (gotta run for the train in a moment), moving the "real work" to index.ts. We removed it a few weeks back because it didn't seem to give us anything we didn't already have.

Re-exporting _everything_ is not going to be practical (tons of stuff) - we only need to export what we need to bootstrap the app, correct? We have, e.g., code which is only imported if environment.dev.ts is built in, otherwise it is not imported at all. If re-exporting everything is required then that will backfire on us (because conditional imports are not syntactically supported by TS, we include dev-mode-only code in environment.dev.

THANK YOU again - i'll get back to you in the next 18-20 hours or so with, i hope, a success report.

**Comment by filipesilva** [comment]
Yeah you just need to re-export the interface to `app`, and since the thing that needs it is `main` then you only need to reexport the imports that `main` needs. Tell me how that goes when you try it again.

Since you have been using the environments functionality, how are you finding it? We're looking to support any number of envs, not just dev and prod, so any feedback you have on that feature would be interesting.

**Comment by sgbeal** [comment]
Briefly from the train: a custom environment was the first thing i tried to work around our broken -prod:  --environment=steve, but it didn't work. We use the env to load a client-side http mocker similar to the in-memory web api bits, and find it very useful. More details tomorrow...

**Comment by sgbeal** [comment]
@filipesilva drum roll, please...

after moving all app-specific imports from main.ts to exports in app/index.ts, and re-importing them in main.ts...

		$ ng serve -prod
		...
		Build successful - 17791ms.
		...

this is 1.0.0-beta.1. The startup XHR count is 18 (down from ~670), thanks to the bundles. Incidentally, i can now throw away the systemjs config bits i did yesterday morning which import angular from its UMD bundles, as that's now obsolete.

If i had the ability to dance, i'd now be performing a Happy Dance!

THANK YOU!!!

Back to the question of custom environment.xxx.ts: i can see a few uses for custom environments, e.g. shops which use dev/integration/production servers with different configurations (commonly seen in java, where properties files are different for each environment, but this is admittedly mostly to provide different backend server names, which isn't so much an issue in JS apps, which tend to communicate only with their origin server). It would be a "nice to have", but probably not a "must have." That said, when i saw the `--environment=prod|dev` flag, i assumed that i could also provide my own.

Another use for environment.xxx.ts might be multi-tenancy, i.e. the ability to provide an app to multiple customers via one code base by swapping out the environment. However, for that to be really useful, custom environments would need the option to turn on bundling or not (e.g. to run in dev mode or prod mode).

**Comment by sgbeal** [comment]
@filipesilva just fyi, there's a concrete example of how we use environment.dev.ts here:

https://github.com/vladvelici/ng2-mock-server/issues/1#issuecomment-218120456

**Comment by filipesilva** [comment]
Nice! Glad it's working now :D

The feedback on the environments kinda confirms the scenario we were going for. The problem with bundling for certain environments is one I'm thinking about solving by decoupling the build from the environment. 

So to speak, you'd be able to do `ng build -prod --env=dev` and conversely `ng build --env=prod`. `--env` would merely specify which environment file to copy over, while `-prod` would bundle stuff/generate service worker manifest/do template compilation.
